Generate Eve's YAML build plan from Python

gdemonet opened the pull request:
## Work In Progress

Plan is to provide tooling for generating the `eve/main.yml` build plan that Eve requires.
Goal is to be faster when adjusting this build plan, and add more complex logic in there that pure YAML would have trouble with.

(dump of the current state of cleaned-up commits, will mark PR as ready once the bulk of `eve/main.yml` is properly rendered)

alexandre-allard committed:
tf: create control plane network & interfaces

Creation of the network used by the control plane,
this allows the usage of a VIP for apiserver which
will be used to test backup/restoration mechanism

alexandre-allard committed:
tf: add script to configure control plane iface

This script creates the configuration of the new
network interface used by the control plane network
and ensures that this interface is up.
It would be better to replace it by a cloud-init
script at some point, but for now, it does the job.

alexandre-allard committed:
tf: add the control plane port to the nodes

Link the control plane interfaces to the nodes

alexandre-allard committed:
tests: fix salt_api tests

Tests were broken because of the introduction of
the VIP, we were taking the first IP listed matching
the CIDR, but VIP will always be the first as this
is the first IP in the subnet.

alexandre-allard committed:
tf: retrieve networks dynamically in bootstrap config

Networks for workload and control plane were hardcoded
in the bootstrap configuration generation script,
we now retrieve them from configured network interfaces
on the host.
Also enable keepalived, since we now have a VIP.

alexandre-allard committed:
tf: add resources to install the bootstrap node

Add the possiblity to deploy metalk8s on the
bootstrap node depending on an optional variable
`metalk8s_iso`, if defined, it'll try to do it.
Usage:
  terraform apply -auto-approve \
    -var='metalk8s_iso=/path/to/your/iso'

It's also possible to specify the destination of
the ISO `metalk8s_iso_dest` and the mountpoint
`metalk8s_iso_mountpoint` on the remote host.

To only deploy a bootstrap node, it's possible
to do:
  terraform apply -auto-approve \
    -var='metalk8s_iso=/path/to/your/iso' \
    -target=null_resource.metalk8s_run_bootstrap

gdemonet committed:
ci: Introduce library to generate CI build plan

In order to ease the management of "eve/main.yml", we want to generate
it from code.

This commit introduces a simple Python library to generate some of the
YAML objects often used in this "eve/main.yml" file. The generation
relies on `OrderedDict` to ensure consistent results across executions.

We also setup unit tests from the start, and provide `tox` entrypoints
for both running those tests and running the buildplan generation (to be
implemented in the future).

gdemonet committed:
ci/buildplan: Add generation of eve/main.yml

Using the Python "lib" in eve/buildplan, we add a Python module, under
eve/buildplan/main.py, and call it from `tox`.
Note that (for now at least), we don't check test coverage for this
`main.py` module.

gdemonet committed:
ci/buildplan: Add steps for pre-merge

This includes all the current "second-level" stages, and a
`SetPropertyFromCommand`.
Core componentes and tests are added as required.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/generate-eve-buildplan` with contents from `improvement/generate-eve-buildplan`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/generate-eve-buildplan origin/development/2.1
 $ git merge origin/improvement/generate-eve-buildplan
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/generate-eve-buildplan
```

gdemonet committed:
ci/buildplan: Group modules into package structure

This eases organising modules and their tests, and keeps the `main.py`
outside of the package itself, that needs to be tested.

Note that to make this work with pytest, we need to add `__init__.py`
files, even if running Python 3.6...

gdemonet committed:
ci/buildplan: Add build stage status automatically

We introduce a `dsl` module, in which we implement the concept of a
mutating stage decorator. The first application is a decorator that
manages creating and uploading a build status at then end of a stage.

Some core steps were added, namely `ShellCommand` and `Upload`, to
achieve this first application.

gdemonet committed:
ci/buildplan: Add steps for "build" stage

We define the `build_all` step command (which simply runs `doit`).
We also introduce two more sets of helpers in the DSL:
- one to manage built artifacts
- one to store common, static, setup steps for later re-use

gdemonet committed:
ci/buildplan: Add steps for "docs" stage

gdemonet committed:
ci/buildplan: Add steps for "lint" stage

gdemonet committed:
ci/buildplan: Add setup steps to OpenStack workers

We add the SSH setup step for OpenStack workers and also include git and
proxy cache setup to both "single-node" and "multiple-nodes" stages.

gdemonet committed:
ci/buildplan: Add steps for "multiple-nodes"

We introduce a large amount of logic in the DSL with this commit:
- stage decorator to automatically manage Terraform deployments from
OpenStack workers
- step and stage decorators to automatically convert local commands into
their SSH equivalents for ease of use with Terraform-orchestrated test
environments

We also split the `buildplan.dsl` module into separate more focused
modules.

gdemonet committed:
lint: Remove Eve generated YAML

Also include all pure YAML files for Salt, not just defaults.yaml
(This meant a couple files changed their extension to ".yaml.j2" to
avoid selecting them with yamllint)

gdemonet committed:
lint: Make sure all Shellscript files are linted

gdemonet committed:
tests/buildplan: Enforce full coverage

alexandre-allard committed:
tf: add bootstrap volume provisioning

Call volume provisioning script in terraform,
if the bootstrap has been run through it.
Step can be disabled adding the following option
on the CLI:
  terraform apply -var=metalk8s_provision_volumes=false

gdemonet committed:
ci/buildplan; Add some debugging build steps

Sometimes, it can be useful to wait in a specific step to leave time for
manual investigations on the worker (or the VMs it spawned).
We add a system that allows:
- to mark a build stage as having a debug step at its end
- to populate a bunch of build properties to enable those debug steps
  based on a single build property called `debug`

See `eve/scripts/mark-for-debug.sh` for more info about the property
format.

alexandre-allard committed:
tf: handle bastion ssh key in terraform

Use terraform provisioner to generate
and copy the bastion SSH keys on every
nodes instead of doing it in the CI.

gdemonet committed:
ci/buildplan: Add steps for "single-node" stage

gdemonet committed:
ci/buildplan: Add step to retrieve artifacts

Instead of manually running `curl` commands, we create a wrapper that
helps retrieve build artifacts (from the current build only, at the
moment).
This allows to define a "naive" `as_remote` implementation, that simply
downloads the artifacts on the worker and copies them through SCP on the
destination host (most of the time, our remote host won't have access to
the artifacts service).

gdemonet committed:
deploy: Extract Terraform tooling from Eve

We extract our Terraform definition files from the multiple-nodes Eve
worker into a set of re-usable, more generic, modules.
This will help both end-users wanting to try MetalK8s on OpenStack and
developers wanting to setup various test environments.
We plan on using it in our CI in a future enhancement, and keep the
existing Terraform definitions for now.

gdemonet committed:
ci/buildplan: Always upload built artifacts

gdemonet committed:
deploy: Show deployment prefix for easier clean-up

In case something went wrong during `terraform destroy`, having a clear
output variable giving the "unique" random prefix that a deployment uses
allows for a simpler manual clean-up procedure.

gdemonet committed:
ci/buildplan: Add helpers to generate ShellCommand

We define two wrappers of `core.ShellCommand` to gather common patterns
into simple arguments.
We also write a bunch of simple helpers to generate shell logic (such as
`_and`, `_or`, or `_for`).

gdemonet committed:
WIP - mount archives first

gdemonet committed:
WIP - deploy Node without relying on hostname

gdemonet committed:
WIP - fix bootstrap preparation script

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/generate-eve-buildplan` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/generate-eve-buildplan`).

```bash
 $ git fetch
 $ git checkout origin/improvement/generate-eve-buildplan
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/generate-eve-buildplan
```

slaperche-scality commented on deploy/README.md:
```diff
@@ -0,0 +1,67 @@
+_Networks_:
+
+- `default_network_name` (default `"tenantnetwork1"`): The default network to
+  use for remote SSH access (from your host and from the Bastion).
```
Is `Bastion` a standard/well-know term? If not, could be worth have a definition somewhere.

slaperche-scality commented on deploy/README.md:
```diff
@@ -0,0 +1,67 @@
+- `default_network_name` (default `"tenantnetwork1"`): The default network to
+  use for remote SSH access (from your host and from the Bastion).
+- `use_private_networks` (default `true`): Create and use two private networks
+  (instead of the default network) for both workload- and control-plane. Values
```
Which `values`? It's a bit ambiguous, is the value of this option or the value for the networks?

slaperche-scality commented on deploy/scripts/create-volumes.sh:
```diff
@@ -0,0 +1,85 @@
+#!/bin/bash
+
+# Note: this script's *defaults* must adhere to the Vagrant environment in
```
Not sure refering to Vagrant is relevant here since it's about deploying a cluster "in the Cloud".

slaperche-scality commented on eve/buildplan/core.py:
```diff
@@ -0,0 +1,162 @@
+    def add(self, stage):
+        existing_stage = next(
+            (s for s in self.stages if s.name == stage.name), None
+        )
```
I wonder if you could use an `OrderedSet` instead of a list here?
(yeah, there is no `OrderedSet`, but an `OrderedMap` with `None` value could do I guess)

slaperche-scality commented on eve/buildplan/core.py:
```diff
@@ -0,0 +1,162 @@
+        self._branches = branches or []
+
+    def __hash__(self):
+        return hash(yamlprint.dump(self.dump()))
```
a bit heavy to do a YAML serialization to compute the hash :/

I understand that you can't hash directly the objects (because `OrderedDict` is not hashable), but maybe using `json.dumps(self.dump(), sort_keys=True)` is a bit less expensive.

slaperche-scality commented on eve/buildplan/core.py:
```diff
@@ -0,0 +1,162 @@
+    """A worker defines how/where to run a stage."""
+
+    @abc.abstractmethod
+    def dump(self):
```
Depends on your version of Python, but:

```
    @abc.abstractmethod
    def dump(self):
        """"here document the interface."""
```
works (and thus, no need for the nocover).

slaperche-scality commented on eve/buildplan/core.py:
```diff
@@ -0,0 +1,162 @@
+    """A step describes what needs to be executed within a stage."""
+
+    # TODO: consider changing halt_on_failure default value to True
+    def __init__(self, name, halt_on_failure=None, always_run=None):
```
Why `None` instead of `False` if `halt_on_failure` is a Boolean?
Ditto for `always_run`.

I guess `None` means `default to whatever is the default in Eve` which may not be very reproducible if Eve ever changes its default values.

slaperche-scality commented on eve/buildplan/test_core.py:
```diff
@@ -0,0 +1,236 @@
+        stage = core.Stage(name="simple", worker=core.LocalWorker(), steps=[])
+
+        project.add(stage)
+        assert sum(stage.name == s.name for s in project.stages) == 1
```
maybe an `assert len(filter(…, project.stages)) == 1` (or using list comprehension) would be more explicit that relying on the fact that `True` is automagically casted to `1`

slaperche-scality commented on eve/buildplan/.coveragerc:
```diff
@@ -1,2 +1,6 @@
 [run]
-omit=*/test_*.py
+omit=
+    # don't cover test files
```
Why? Would allow to catch dead code in tests, no?

slaperche-scality commented on eve/buildplan/main.py:
```diff
@@ -0,0 +1,36 @@
+"""Generate the Eve YAML description of build plan for MetalK8s."""
```
Shouldn't you have a shebang to at least clarify if you're using Python 2 or 3?

slaperche-scality commented on eve/buildplan/main.py:
```diff
@@ -0,0 +1,36 @@
+
+if __name__ == "__main__":
+    build_plan = build_project().dump()
+    yamlprint.dump(build_plan, stream=sys.stdout)
```
instead of hardcoding `stdout`, take a filename as input from the CLI (default to `stdout` if not provided if you want)
That way it's easier to test, and you don't need the `bash -c` in tox (because you no longer need shell redirection).

slaperche-scality commented on eve/buildplan/main.py:
```diff
@@ -58,7 +60,16 @@ def build():
-        steps=[],
+        steps=[
+            build_all(),
+            *dsl.copy_artifacts(
+                [
+                    "_build/metalk8s.iso",
+                    "_build/SHA256SUM",
+                    "_build/root/product.txt",
+                ]
+            ),
```
Couldn't this be handled by the `WithArtifacts` decorator?

slaperche-scality commented on eve/main.yml:
```diff
@@ -82,7 +82,7 @@ stages:
     - Upload:
         name: Upload artifacts
-        source: artifacts
+        source: &id002 artifacts
```
minor: `id002`? Cant' we generate better name (using `name` above`)?

slaperche-scality commented on eve/buildplan/buildplan/shell.py:
```diff
@@ -0,0 +1,84 @@
+
+# Helpers {{{
+def _fmt_args(*args, join_with=" "):
+    """Join an iterable of `Optional[str]`, omitting `None`s."""
```
Any reason why `None` instead of falsy value (i.e. None and empty string)?

slaperche-scality commented on eve/buildplan/buildplan/dsl/artifacts.py:
```diff
@@ -0,0 +1,79 @@
+from collections import Mapping
```
I would have done the split into a dedicated commit, because here it's hard to tell the difference between the new code that we should review and the old code that is simply moved around.

slaperche-scality commented on tox.ini:
```diff
@@ -88,11 +88,11 @@ commands =
     yamllint==1.15.0
 commands =
-    bash -c "yamllint eve/main.yml salt/metalk8s/defaults.yaml"
+    bash -c "yamllint salt"
```
Why aren't we linting the `eve/main.yaml`?

slaperche-scality commented on eve/buildplan/buildplan/core.py:
```diff
@@ -260,6 +260,7 @@ class ShellCommand(Step):
         workdir="workdir",
         sigterm_time="sigtermTime",
         use_pty="usePTY",
+        timeout="timeout",
```
`ci/buildplan; ` => should be `:` (in the commit message)

slaperche-scality commented on eve/buildplan/buildplan/core.py:
```diff
@@ -273,10 +274,8 @@ def dump_arguments(self):
-        Step.OPTIONAL_ARGS,
-        urls="urls",
-    )
+    OPTIONAL_ARGS = dict(Step.OPTIONAL_ARGS, urls="urls")
```
unrelated change :stuck_out_tongue:
